### PR TITLE
First steps towards pythonizing cam buildnml scripts

### PR DESCRIPTION
**Pull request by jgfouca**
Pythonize the top-level buildnml script.

[BFB]

### CODE REVIEW

**Comment by jgfouca** [comment]
I've tested this PR and it produces identical nml files to master for all of e3sm_integration.

File: components/cam/cime_config/buildnml [hunk 1]
```diff
@@ -1,225 +1,245 @@
...
+    if "-01-01" in run_startdate or "-09-01" in run_startdate:
+        ignore = "-ignore_ic_year"
+
+    if cam_nml_use_case not in [None, "UNSET"]:
+        usecase = "-use_case " + cam_nml_use_case
+    else:
+        usecase = " "
+
+    inst_string = ""
+    for inst_counter in xrange(1, ninst_atm + 1):
```
Comment (mfdeakin-sandia) [inline components/cam/cime_config/buildnml]: Python 3 compatibility for `xrange`: http://python-future.org/compatible_idioms.html#xrange
Or just use `range`

Comment (jgfouca) [reply components/cam/cime_config/buildnml]: Since this is E3SM, not CIME, I don't think we need to worry about py3 for now. @rljacob , what do you think?

Comment (mfdeakin-sandia) [reply components/cam/cime_config/buildnml]: That's fair. I just didn't see anything else obvious (though looking again I see potential division issues) blocking python3 support so thought I'd mention it

**Comment by jgfouca** [comment]
Waiting for reviews from @singhbalwinder and @wlin7

**Comment by singhbalwinder** [comment]
Based on my limited python knowledge, it looks like it covers all the needed functionality. Your tests also indicate that there is no loss of functionality, so I am going to approve this. Thanks for working on this task. 

Did we get any performance benefit as well, that is, is it faster than Perl?

**Comment by jgfouca** [comment]
@singhbalwinder , there will be some performance benefit. All the xmlqueries that got replaced at the top of this script will make it faster. The main benefits will come when the 4000-line perl scripts called by this script are converted.

**Comment by jgfouca** [comment]
Merged to next.

**Comment by mfdeakin-sandia** [comment]
I was curious, so here's the performance difference:
Python
real	0m0.712s

Perl
real	0m14.363s

ie, 20x speedup. Thanks Jim!

**Comment by jgfouca** [comment]
@mfdeakin-sandia thanks for doing that! Wow, that's a much bigger speedup than I thought. Did you isolate the performance of buildnml or does this include calls to perl?

**Comment by mfdeakin-sandia** [comment]
I didn't isolate the performance benefit, but I don't think the perl script does much in the way of XML, and what it does do (using XML::Lite) is probably faster than starting a new Python instance each change

**Pull request merged**